src/Symfony/Component/DependencyInjection/LazyProxy/PhpDumper/LazyServiceDumper.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\LazyProxy\PhpDumper;
+
+use Symfony\Bridge\ProxyManager\LazyProxy\PhpDumper\ProxyDumper;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Exception\LogicException;
+use Symfony\Component\VarExporter\LazyGhostObjectInterface;
+use Symfony\Component\VarExporter\LazyGhostObjectTrait;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class LazyServiceDumper implements DumperInterface
+{
+    public function __construct(
+        private string $salt = '',
+    ) {
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isProxyCandidate(Definition $definition, bool &$asGhostObject = null): bool
+    {
+        $asGhostObject = false;
+
+        if ($definition->hasTag('proxy')) {
+            if (!$definition->isLazy()) {
+                throw new InvalidArgumentException(sprintf('Invalid definition for service of class "%s": setting the "proxy" tag on a service requires it to be "lazy".', $definition->getClass()));
+            }
+
+            return true;
+        }
+
+        if (!$definition->isLazy()) {
+            return false;
+        }
+
+        if (!($class = $definition->getClass()) || !(class_exists($class) || interface_exists($class, false))) {
+            return false;
+        }
+
+        $class = new \ReflectionClass($class);
+
+        if ($asGhostObject = !$class->isAbstract() && !$class->isInterface() && ('stdClass' === $class->name || !$class->isInternal())) {
```

Comment:
what about final classes ? shouldn't we have a clean error handling ?

Reply:
I added a check to throw an exception for final classes.